src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -92,10 +117,33 @@ int MPIDI_XPMEM_mpi_finalize_hook(void)
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_FINALIZE_HOOK);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_FINALIZE_HOOK);
 
+    if (MPIR_Process.comm_world->node_comm) {
+        MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+        mpi_errno = MPIR_Barrier(MPIR_Process.comm_world->node_comm, &errflag);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+    /* Free temporary shared buffer */
+    for (i = 0; i < MPIDI_XPMEM_global.num_local; ++i) {
+        MPIDI_XPMEM_seg_deregist(MPIDI_XPMEM_global.coop_counter_origin_direct[i]);
+    }
+
+    MPL_free(MPIDI_XPMEM_global.coop_counter_direct);
+    MPL_free(MPIDI_XPMEM_global.coop_counter_origin_direct);
+
+    /* free all queued counter obj */
+    while (MPIDI_XPMEM_global.cnt_free_queue) {
+        MPIDI_XPMEM_cnt_free_obj_t *head = MPIDI_XPMEM_global.cnt_free_queue;
+        DL_DELETE(MPIDI_XPMEM_global.cnt_free_queue, head);
+        MPIR_Handle_obj_free(&MPIDI_XPMEM_cnt_mem, (void *) head);
+        MPL_free(head);
+    }
```

Comment:
It is incorrect if any of the queued counters are not freed because it means the corresponding receive requests are not completed. I'd suggest you add `MPIR_Assert` to check the flag.

Besides, should `MPIR_Handle_obj_free` free `head->counter_ptr`? Can you please write test program to check the counter free queue code is correct? The MPICH test suite may not cover this case.